adept/container/actorlearner/learner_container.py

```diff
@@ -0,0 +1,374 @@
+            self.optimizer = optim_fn(self.network.parameters())
+
+        # LEARNER / EXP
+        rwd_norm = REGISTRY.lookup_reward_normalizer(
+            args.rwd_norm).from_args(args)
+        actor_cls = REGISTRY.lookup_actor(args.actor_host)
+        builder = actor_cls.exp_spec_builder(
+            env.observation_space,
+            env.action_space,
+            net.internal_space(),
+            args.nb_env * args.nb_learn_batch
+        )
+        w_builder = REGISTRY.lookup_actor(args.actor_worker).exp_spec_builder(
+            env.observation_space,
+            env.action_space,
+            net.internal_space(),
+            args.nb_env
+        )
+        actor = actor_cls.from_args(args, env.action_space)
+        learner = REGISTRY.lookup_learner(args.learner).from_args(args, rwd_norm)
+
+        exp_cls = REGISTRY.lookup_exp(args.exp).from_args(args, builder)
+
+        self.actor = actor
+        self.learner = learner
+        self.exp = exp_cls.from_args(args, builder).to(device)
+
+        # Rank 0 setup, load network/optimizer and create SummaryWriter/Saver
+        if rank == 0:
+            if args.load_network:
+                self.network = self.load_network(self.network, args.load_network)
+                logger.info('Reloaded network from {}'.format(args.load_network))
+            if args.load_optim:
+                self.optimizer = self.load_optim(self.optimizer, args.load_optim)
+                logger.info('Reloaded optimizer from {}'.format(args.load_optim))
+
+            logger.info('Network parameters: ' + str(self.count_parameters(net)))
+            self.summary_writer = SummaryWriter(log_id_dir)
+            self.saver = SimpleModelSaver(log_id_dir)
+
+    def run(self, workers, profile=False):
+        if profile:
+            try:
+                from pyinstrument import Profiler
+            except:
+                raise ImportError('You must install pyinstrument to use profiling.')
+            profiler = Profiler()
+            profiler.start()
+
+        # setup queuer
+        self.rollout_queuer = RolloutQueuerAsync(workers, self.nb_learn_batch, self.rollout_queue_size)
+        self.rollout_queuer.start()
+
+        # initial setup
+        global_step_count = self.initial_step_count
+        next_save = self.init_next_save(self.initial_step_count, self.epoch_len)
+        prev_step_t = time()
+        ep_rewards = torch.zeros(self.nb_env)
+        start_time = time()
+
+        # loop until total number steps
+        print('{} starting training'.format(self.rank))
+        while not self.done(global_step_count):
+            self.exp.clear()
+            # Get batch from queue
+            rollouts, terminal_rewards, terminal_infos = self.rollout_queuer.get()
+
+            # Iterate forward on batch
+            self.exp.write_exps(rollouts)
+            # keep a copy of terminals on the cpu it's faster
+            rollout_terminals = torch.stack(self.exp['terminals']).numpy()
```

Comment:
neat, didn't know .numpy() automatically called .cpu() for you

Reply:
it doesn't actually, rollouts are from ray so they're on the cpu already.